docs/lofi-hifi-backends.md

```diff
@@ -0,0 +1,279 @@
+
+```console
+ilab
+|
+|_______model
+|       |
+|       |____convert
+|       |____download
+|       |____train (--convert)
+|       |    |
+|       |    |______peft *
+|       |    |______phased *
+|       |
+|       |____serve   (-i)
+|       |____chat
+|       |____inference
+|       |
+|_______data
+|       |____generate
+|       |    |
+|       |    |______lofi * (name pending)
+|       |    |______hifi * (name pending)
+|       |
+|       |____mix *
+|       |
+|       |
+|_______checkpoint
+|       |
+|       |____evaluate *
+|       |
+|_______config
+|       |
+|       |____init
+|_______tax
+|       |
+|       |____diff
+|       |____check
+|       |____download
+```
+
+
+The starred commands are the new ones under the redesigned structure. We are opting to add peft and phased as commands under train and lofi/hifi under generate rather than flags given their completely different functionality and backend structure.
+
+`ilab train peft` would utilize much of the existing upstream training code with the following differences:
+
+- deepseed support
+- fsdp support (eventually)
+- full evaluation at the end of training
+- inter-epoch evaluation using pytorch's built in evaluators that fit for value loss primarily.
+
+`ilab train phased` would utlize much of the backend training code. Primarily doing the following:
+
+- running a single phase of training
+- outputting checkpoint dirs for evaluation to be run on
+- deepspeed support
+- fsdp support.
+
+`ilab train phased` will be used in conjunction with `ilab checkpoint evaluate`. This command will run full scale inter-checkpoint evaluation on the given directory. An output dir will then hold then best checkpoint and all necessary data to run another `ilab train phased` command on.
+
+Much of the code for `ilab train phased` and `ilab train peft` will live in the CLI, and hopefully will be shared. However, plans to move much of this code to an external training library should eventually happen. This is most likely not in scope for the first version of this work.
+
+### Reasoning
+
+Plugging into hardware acceleration and multi-phase training is the logical next step for ilab. Ensuring we do this in a clean way that does not overload our current commands is also crucial. Many of the processes in the backend are confusing so we want to abstract some of the steps away from users while also giving them a reasonable amount of choice in configuring these new processes. However, maintaining the current laptop story is important to users without hardware access. Splitting these two paths into separate commands maintains the integrity of each.
+
+Not only maintaining the current laptop story but enhancing it. Qlora by itself is great, but having LoRA based training with the ability to run ZeRO phaes 1-3 or full FSDP accleration given the right hardware will improve the experience for people on 1 or 2 GPU systems (which are quite common).
```

Comment:
```suggestion
Not only maintaining the current laptop story but enhancing it. Qlora by itself is great, but having LoRA based training with the ability to run ZeRO phases 1-3 or full FSDP acceleration given the right hardware will improve the experience for people on 1 or 2 GPU systems.
```

Reply:
How do we know that "people on 1 or 2 GPU systems" is common? If we can't easily prove it, I'd suggest to remove it. It's not adding much to your point that the laptop story is important to maintain and enhance.